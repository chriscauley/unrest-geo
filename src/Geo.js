import { range } from 'lodash'

import { mod } from '../utils'
import Look from './Look'

const geo_cache = {}

class Geo {
  constructor(x0, x_max, y0, y_max) {
    if (x_max === undefined) {
      x_max = x0
    }
    if (y0 === undefined) {
      y_max = x_max - 1
      x_max = x0 - 1
      x0 = 0
      y0 = 0
    }

    const W = Math.abs(x_max - x0) + 1
    const H = Math.abs(y_max - y0) + 1

    // constants
    Object.assign(this, { x0, y0, W, H })

    // cache tables
    Object.assign(this, {
      xys: [],
      indexes: [],
      AREA: W * H,
      _dindex_names: ['u', 'l', 'r', 'd'],
      _dindex_chars: ['^', '<', '>', 'v'],
      _dindex2char: {},
      _name2dindex: {},
      _dindex2name: {},
      dindexes: [-W, -1, 1, W], // u, l, r, d
      rot_dindexes: {
        [-W]: [-W, -1, 1, W], // u, l, r, d
        [W]: [W, 1, -1, -W], // d, l, r, u
        [1]: [1, -W, W, -1], // r, u, d, l
        [-1]: [-1, W, -W, 1], // l, d, u, r
      },
    })

    this.dindexes.forEach((dindex, i) => {
      const name = this._dindex_names[i]
      this._dindex2name[dindex] = name
      this._name2dindex[name] = dindex
      this._dindex2char[dindex] = this._dindex_chars[i]
    })

    this.CENTER = this.xy2index([
      Math.floor((this.x0 + this.W) / 2),
      Math.floor((this.y0 + this.H) / 2),
    ])

    range(y0, y_max + 1).forEach((y) =>
      range(x0, x_max + 1).forEach((x) => {
        const xy = [x, y]
        this.xys.push(xy)
        this.indexes.push(this.xy2index(xy))
      }),
    )

    this.look = Look(this)
  }

  index2xy = (i) => [mod(i, this.W), Math.floor(i / this.W)]
  xy2index = (xy) => mod(xy[0] + xy[1] * this.W, this.AREA)
  print(board, options = {}) {
    const {
      from_xy = [this.x0, this.y0],
      to_xy = [this.x0 + this.W-1, this.y0 + this.H-1],
      delimiter = '',
      empty = ' ',
      extras,
      title,
    } = options
    const xs = range(from_xy[0], to_xy[0] + 1)
    const ys = range(from_xy[1], to_xy[1] + 1)
    const lines = ys.map((y) =>
      xs
        .map((x) => board[this.xy2index([x, y])])
        .map((s) => (s === undefined ? empty : s))
        .join(delimiter),
    )
    if (extras) {
      extras.forEach((e, i) => (lines[i] += '\t' + e))
    }
    title && lines.unshift(title)
    return lines.join('\n')
  }
  inBounds(xy) {
    return (
      xy[0] >= this.x0 && xy[0] < this.x0 + this.W && xy[1] >= this.y0 && xy[1] < this.y0 + this.H
    )
  }
  slice(xy, W, H) {
    const out = []
    const ys = range(xy[1], xy[1] + H)
    const xs = range(xy[0], xy[0] + W)
    ys.forEach((y) => xs.forEach((x) => out.push(this.xy2index([x, y]))))
    return out
  }
  floorDindex(dindex) {
    // if dindex is a multiple of W it is in thy y direction
    // otherwise it is in the x direction
    return (dindex % this.W === 0 ? this.W : 1) * Math.sign(dindex)
  }
}

export default (x0, x_max, y0, y_max) => {
  // since geo's are meant to be deterministic with the above parameters they can be cached
  const key = `${x0},${x_max},${y0},${y_max}`
  geo_cache[key] = geo_cache[key] || new Geo(x0, x_max, y0, y_max)
  return geo_cache[key]
}
