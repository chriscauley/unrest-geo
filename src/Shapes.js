import { range } from 'lodash'

const Shapes = (geo) => ({
  circle: (dist, dindex) => {
    const out = []
    const o_dindex = geo.rot_dindexes[dindex][1]
    out.push(dindex * dist)
    range(1, dist + 1).forEach((i) => {
      out.push(dindex * (dist - i) - o_dindex * i)
      out.push(dindex * (dist - i) + o_dindex * i)
    })
    range(1, dist).forEach((i) => {
      out.push(-dindex * (dist - i) - o_dindex * i)
      out.push(-dindex * (dist - i) + o_dindex * i)
    })
    out.push(dindex * -dist)
    return out
  },
  box: (dist, dindex) => {
    const out = []
    const o_dindex = geo.rot_dindexes[dindex][1]

    // top row
    range(-dist, dist + 1).forEach((i) => out.push(dist * dindex - i * o_dindex))

    // left and right sides in middle
    range(1, 2 * dist).forEach((i) => {
      out.push((dist - i) * dindex + o_dindex * dist)
      out.push((dist - i) * dindex - o_dindex * dist)
    })

    // bottom row
    range(-dist, dist + 1).forEach((i) => out.push(-dist * dindex - i * o_dindex))
    return out
  },
  three: (dist, dindex) => {
    const target = dist * dindex
    const o_dindex = geo.rot_dindexes[dindex][1]
    return [target, target + o_dindex, target - o_dindex]
  },
  cone: (dist, dindex) => {
    const target = dist * dindex
    const out = [target]
    const o_dindex = geo.rot_dindexes[dindex][1]
    range(1, dist).forEach((o_dist) => {
      out.push(target + o_dist * o_dindex)
      out.push(target - o_dist * o_dindex)
    })
    return out
  },
  f: (dist, dindex) => [dist * dindex],
  lr: (dist, dindex) => {
    const o_dindex = geo.rot_dindexes[dindex][1]
    return [o_dindex * dist, -o_dindex * dist]
  },
  cross: (dist, dindex) => geo.rot_dindexes[dindex].map((dindex2) => dist * dindex2),
})

Shapes.list = []

Object.keys(Shapes()).forEach((shape) => {
  Shapes.list.push('__' + shape)
  Shapes.list.push(shape)
})

export default Shapes
