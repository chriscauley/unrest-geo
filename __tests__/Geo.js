import Geo from '../src/Geo'
import Shapes from '../src/Shapes'

const snapIndexes = (geo, indexes, title) => {
  if (process.env.VERBOSE) {
    const board = {}
    board[geo.CENTER] = 'X'
    indexes.forEach((index, i) => (board[index] = numalpha[i] || '?'))
    console.log(geo.print(board, { empty: '.', title })) // eslint-disable-line
  }
  expect(indexes.join(',')).toMatchSnapshot()
}

test('xy2index', () => {
  const geo = Geo(5, 5)
  const b = { geo }
  const set = (xy, v) => (b[geo.xy2index(xy)] = v)
  set([2, 2], 1)
  set([-1, 0], 2)
  set([3, -2], 3)
  set([1, 8], 4)
  set([5, 3], 5)
  set([6, 0], 6)
  snapIndexes(geo, Object.keys(b), 'xy2index board')
})

const numalpha = '0123456789abcdefghijklmnopqrstuvqxyz'

Shapes.list
  .filter((s) => !s.startsWith('__'))
  .forEach((shape) => {
    test('look.' + shape, () => {
      const dist = 3
      const S = dist * 2 + 1
      const geo = Geo(S, S)
      const dindex = 1
      const indexes = geo.look(shape, geo.CENTER, dist, dindex)
      const title = `${shape} ${dist} ${geo._dindex2name[dindex]}`
      snapIndexes(geo, indexes, title)
    })
  })

test('`geo.look(shape, index, dist, dindex)` changes direction with dindex', () => {
  const shape = 'cone'
  const dist = 4
  const S = dist * 2 + 1
  const geo = Geo(S, S)
  const index = geo.CENTER
  geo.dindexes.forEach((dindex) => {
    const indexes = geo.look(shape, index, dist, dindex)
    const dir = geo._dindex2name[dindex]
    snapIndexes(geo, indexes, `"${shape}" facing ${dir} at dist ${dist}`)
  })
})

test('non-zero x0, y0', () => {
  const geo = new Geo(5, 5, { x0: -1, y0: -1 })
  const b = {}
  const set = (xy, v) => (b[geo.xy2index(xy)] = v)
  set([0, 0], 0)
  set([1, 1], 1)
  set([1, 2], 2)
  set([-1, -1], '-')
  expect(JSON.stringify(b)).toMatchSnapshot()
})
