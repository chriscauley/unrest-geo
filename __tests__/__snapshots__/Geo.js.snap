// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`geo.look(shape, index, dist, dindex)\` changes direction with dindex 1`] = `"31,22,21,23,13,12,14,11,15,4,3,5,2,6,1,7"`;

exports[`\`geo.look(shape, index, dist, dindex)\` changes direction with dindex 2`] = `"39,38,47,29,37,46,28,55,19,36,45,27,54,18,63,9"`;

exports[`\`geo.look(shape, index, dist, dindex)\` changes direction with dindex 3`] = `"41,42,33,51,43,34,52,25,61,44,35,53,26,62,17,71"`;

exports[`\`geo.look(shape, index, dist, dindex)\` changes direction with dindex 4`] = `"49,58,59,57,67,68,66,69,65,76,77,75,78,74,79,73"`;

exports[`look.box 1`] = `"18,25,32,17,31,16,23,30,12,19,26,33,40,11,39,10,38,9,37,8,15,22,29,36,6,13,20,27,34,41,48,5,47,4,46,3,45,2,44,1,43,0,7,14,21,28,35,42"`;

exports[`look.circle 1`] = `"25,31,17,23,26,32,18,38,10,30,16,22,27,33,19,39,11,45,3,29,15,37,9,21"`;

exports[`look.cone 1`] = `"25,26,19,33,27,20,34,13,41"`;

exports[`look.cross 1`] = `"25,17,31,23,26,10,38,22,27,3,45,21"`;

exports[`look.f 1`] = `"25,26,27"`;

exports[`look.lr 1`] = `"17,31,10,38,3,45"`;

exports[`look.three 1`] = `"25,18,32,26,19,33,27,20,34"`;

exports[`xy2index 1`] = `"6,12,16,18,20,24,geo"`;
